### ISSUE TITLE
Array elements can't be validated

### ISSUE THREAD

**Issue by bjansen** [jakartaee/validation-spec#282]
Hi,

I'm trying to validate each element of an array, rather than the size of said array.

According to [jls§9.7.4](https://docs.oracle.com/javase/specs/jls/se22/html/jls-9.html#jls-9.7.4), annotations can be used on arrays like this:

<blockquote>

Type annotations can apply to an array type or any component type thereof ([§10.1](https://docs.oracle.com/javase/specs/jls/se22/html/jls-10.html#jls-10.1)). For example, assuming that A, B, and C are annotation interfaces meta-annotated with `@Target(ElementType.TYPE_USE)`, then given the field declaration:

<pre>
@C int @A [] @B [] f;
</pre>

`@A` applies to the array type int[][], `@B` applies to its component type int[], and `@C` applies to the element type int. For more examples, see [§10.2](https://docs.oracle.com/javase/specs/jls/se22/html/jls-10.html#jls-10.2).
</blockquote>

So essentially, these two declarations are "equivalent" according to the JLS:

```java
private @NotEmpty List<@Size(min=1,max=3) String> myList;

private @Size(min=1,max=3) String @NotEmpty[] myArray;
```

It seems to me that currently it is only possible to validate the length of an array, but not validate each element inside this array. I also have the impression that the current behavior is different from what is described in the above JLS excerpt, and also different from how `List`s are validated:
* `@Size(min = 1, max = 3) String[]` validates the array length whereas it should in theory validate each `String` element
* `String @Size(min = 1, max = 3)[]` should validate the array length but it does nothing at the moment.

See this sample project for JUnit test cases to reproduce the problem: https://github.com/bjansen/arrays-validation

**Comment by marko-bekhta** [comment]
Hey @bjansen ,

To comply with JLS rules for arrays, we will be breaking backwards compatibility, and may also end up with ambiguity for constraint declarations: for `private @Size(min=1,max=3) String[] myArray`  `@Size`  will be considered both field and element type.

For more details, see:
- this mailing list discussion: https://lists.jboss.org/pipermail/beanvalidation-dev/2017-April/001274.html
- jira issue to deal with arrays in the Hibernate Validator: https://hibernate.atlassian.net/browse/HV-1428

**Comment by bjansen** [comment]
I'm not sure I understand correctly, do you mean that this will not be fixed in the near future because it would break existing code?

**Comment by marko-bekhta** [comment]
No, not necessarily. My comment was more to link to the reasons why we are where we are right now. Also, you see, those threads are quite old (since the BV 2.0 times). And we (at Hibernate Validator) haven't heard much from the users asking for any progress on this problem, hence it wasn't a priority to look into it.

On those message threads, there were some ideas on how to possibly make this work, but we'd have to start with those experiments in the impl first before we can suggest something to be set in the spec.
